tools/cli/src/helpers/common.ts

```diff
@@ -1,42 +1,114 @@
 import {
     Asset,
     AssetWithContent,
-    AsyncResult,
+    Err,
+    Ok,
+    R,
+    RawSourceMap,
     ResultPromise,
+    SourceAndSourceMap,
     SourceProcessor,
     loadSourceMap,
+    parseJSON,
+    pipe,
     readFile,
+    writeFile,
 } from '@backtrace-labs/sourcemap-tools';
 
 export function toAsset(file: string): Asset {
     return { name: file, path: file };
 }
 
-export function readAsset(asset: Asset): ResultPromise<AssetWithContent<string>, string> {
-    return AsyncResult.equip(readFile(asset.path)).then((content) => ({ ...asset, content })).inner;
+export function readSource<T extends Asset>(asset: T): ResultPromise<AssetWithContent<string>, string> {
+    return pipe(
+        asset.path,
+        readFile,
+        R.map((content) => ({ ...asset, content })),
+    );
 }
 
-export function loadSourceMapFromPathOrFromSource(sourceProcessor: SourceProcessor) {
-    return function loadSourceMapFromPathOrFromSource(asset: Asset) {
-        return AsyncResult.fromValue<Asset, string>(asset)
-            .then(loadSourceMap)
-            .thenErr(
-                () =>
-                    AsyncResult.fromValue<Asset, string>(asset)
-                        .then(resolveSourceMapPathFromSource(sourceProcessor))
-                        .then(loadSourceMap).inner,
-            ).inner;
+export function readSourceMap<T extends Asset>(asset: T): ResultPromise<AssetWithContent<RawSourceMap>, string> {
+    return pipe(
+        asset.path,
+        readFile,
+        R.map(parseJSON<RawSourceMap>),
+        R.map((content) => ({ ...asset, content })),
+    );
+}
+
+export function readSourceAndSourceMap(sourceProcessor: SourceProcessor) {
+    return function _readSourceAndSourceMap(sourceAsset: Asset): ResultPromise<SourceAndSourceMap, string> {
+        return pipe(
+            sourceAsset,
+            readSource,
+            R.map((source) =>
+                pipe(
+                    source.content,
+                    (content) => sourceProcessor.getSourceMapPathFromSource(content, sourceAsset.path),
+                    R.map((path) => ({ name: path, path } as Asset)),
+                    R.map(readSourceMap),
+                    R.map((sourceMap) => ({ source, sourceMap } as SourceAndSourceMap)),
+                ),
+            ),
+        );
+    };
+}
+
+export function writeAsset<T extends AssetWithContent<unknown>>(asset: T) {
+    return pipe(
+        asset.content,
+        (content) => (typeof content === 'object' ? JSON.stringify(content) : String(content)),
```

Comment:
I think you should use toString instead of String. 
```suggestion
        (content) => (typeof content === 'object' ? JSON.stringify(content) : content?.toString() ?? "",
```

Reply:
Typescript is erroring here that `.toString()` is not defined.